File: libs/core/src/symbols/variable_symbol.cpp
/**
 * @file variable_symbol.cpp
 * @author Francesco Corso
 * @date 8/8/2026
 * @brief
 */
module prism.core:symbols.variable_symbol.impl;

import :symbols.variable_symbol;
import :symbols.type_symbol;

namespace prism
{
    void VariableSymbol::write_display_string(TextWriter &writer) const
    {
        writer.write(name().as_string_view());
        writer.write(": ");
        type().write_display_string(writer);
    }
} // namespace prism

File: src/Prism.Core/Symbols/VariableSymbol.cs
using System.Diagnostics.CodeAnalysis;
using Prism.Core.Semantic;

namespace Prism.Core.Symbols;

public closed class VariableSymbol : ValueSymbol
{
    private protected VariableSymbol(string name, Symbol? containingSymbol)
        : base(name, containingSymbol) { }

    public abstract bool IsGlobal { get; }

    public bool IsLocal => !IsGlobal;

    public abstract bool HasInitializer { get; }

    [MemberNotNullWhen(true, nameof(ConstantValue))]
    public bool IsConst => ConstantValue is not null;

    public abstract ConstantValue? ConstantValue { get; }

    public sealed override void WriteDisplayString(TextWriter writer)
    {
        writer.Write(Name);
        writer.Write(": ");
        Type.WriteDisplayString(writer);
    }
}
